File: Graphs/BFS.cpp
#include <vector>
#include <limits>
#include <queue>

// adjacency matrix version
void BFS(std::vector<std::vector<int>> graph, int s){
	int num_ver = graph.size();
	std::vector<bool> explored_vector (num_ver, false);
	std::vector<int> length_vector (num_ver, std::numeric_limits<int>::max());
	explored_vector[s-1] = true;
	length_vector[s-1] = 0;
	std::queue<int> q;
	q.push(s);
	while(!q.empty()){
		int v = q.front();
		q.pop();
		for (int i = 0; i < num_ver; i++){
			if (graph[v-1][i] == 1){
				if (explored_vector[i] == false){
					explored_vector[i] = true;
					length_vector[i] = length_vector[v-1] + 1;
					q.push(i+1);
				}	
			}
		}
	}
	return ;
}

//adjacency list version
void al_BFS(std::vector<int> vertices, std::vector<int> edges, int s){
	int num_vertices = vertices.size(); 
	std::vector<bool> explored(num_vertices, false); // list that represents whether a vertex i at index i-1 has been explored
	explored[s-1] = true; // starting vertex is explored
	std::queue<int> q;
	q.push(vertices[s-1]);
	while (!q.empty()){
		int v = q.front();
		q.pop();
		for (int i = vertices[v]; i < vertices[v+1]; i++){ //for all edges from v
			if (!explored[edges[i]]){ // if not explored, explore then add to queue
				explored[edges[i]] = true;
				q.push(edges[i]);
			}
		}
	}
	return ;
}

std::vector<int> UCC(std::vector<std::vector<int>> graph){
	int num_ver = graph.size();
	int num_cc = 0;
	std::vector<int> cc_vector(num_ver,0);
	for (int i = 0; i < num_ver; i++){
		if (cc_vector[i] == 0){
			num_cc++;
			cc_vector[i] = num_cc;
			std::queue<int> q;
			q.push(i+1);
			while(!q.empty()){
				int v = q.front();
				q.pop();
				for (int j = 0; j < num_ver; j++){
					if (graph[v-1][j] == 1){
						q.push(j+1);
						cc_vector[j] = num_cc;
					}
				}
			}
		}
	}
	return cc_vector;
}

std::vector<int> al_UCC(std::vector<int> vertices, std::vector<int> edges, int s){
	int num_ver = vertices.size();
	int num_cc = 0;
	std::vector<int> cc_vector(num_ver,0);
	for (int i = 0; i < num_ver; i++){
		if (cc_vector[i] == 0){
			num_cc++;
			cc_vector[i] = num_cc;
			std::queue<int> q;
			q.push(i+1);
			while(!q.empty()){
				int v = q.front();
				q.pop();
				for (int j = vertices[v]; j < vertices[v+1]; j++){
					int voi = edges[j];
					if(cc_vector[voi] == 0){				
						q.push(voi);
						cc_vector[voi] = num_cc;
					}
				}
			}
		}
	}
	return cc_vector;
}

File: Graphs/BFS.py
import sys
import queue
'''
	Given G(V,E) in adjancency matrix, and a vertex s, with all edges length == 1, 
	return shortest length from s to some vertex v for all vertex v in V.
'''
def BFS(adj_mat, s):
	num_vertex = len(adj_mat) # number of vertices 
	explored_list = [False]*num_vertex # boolean list of whether the ith vertex has been explored
	length_list = [sys.maxsize]*num_vertex # length list of distance from s to the ith vertex
	explored_list[s-1] = True # initialize so that s is explored
	length_list[s-1] = 0 # initialize so that s is 0 away from itself
	q = queue.SimpleQueue() # initialize queue
	q.put(s) # and put s
	while not q.empty(): # while there exists a vertex to travel to
		v = q.get() # pop a vertex
		edges = adj_mat[v-1] # get the according row of adjancency matrix 
		for c in range(num_vertex): 
			if c != 0: # find the edge
				if explored_list[c] == False: #if not explored
					explored_list[c] = True # now explored
					length_list[c] = length_list[v-1] + 1 # assign length
					q.put(c+1) # put it in the queue!
	return length_list

'''
	Given G(V,E) in adjancency matrix, and a vertex s, with all edges length == 1, 
	return to which connected component a vertex v belongs to for all v in V.
'''
def UCC(adj_mat):
	num_vertices = len(adj_mat)
	cc_list = [0] * num_vertices
	num_CC = 0
	for i in range(num_vertices): # all vertices
		if cc_list[i] == 0: # if not explored
			num_CC += 1
			cc_list[i] = num_CC
			q = queue.SimpleQueue() # initialize queue
			q.put(s) # and put s
			while not q.empty(): # while there exists a vertex to travel to
				v = q.get() # pop a vertex
				edges = adj_mat[v-1] # get the according row of adjancency matrix 
				for c in range(num_vertices): 
					if c != 0: # find the edge
						if cc_list[c] == 0: #if not explored
							cc_list[c] = num_CC # now explored
							q.put(c+1) # put it in the queue!
	return cc_list










	